Fix 16191 - std/digest/digest.d should be renamed to package.d

wilzbach opened the pull request:
CC @schveiguy @jpf91 

From Bugzilla:

> The fully qualified name of the digest function template (https://dlang.org/phobos/std_digest_digest.html#digest) is std.digest.digest.digest. This is because std.digest is a package, with a module named digest in it, and the function digest inside that.

> I think it's reasonable to have std.digest.digest. Even though there's repetition, it's clear that the first digest is a module, the second is a function name. But there's no reason (especially with the advent of package.d) to have a std.digest.digest module.

I heard somewhere that `std/digest/digest` comes from a time where `package.d` wasn't supported yet.
The rename is a separate commit, s.t. git detects it correctly and can retain the history.

WebFreak001 commented:
removing in November 2017 already? :/

wilzbach commented:
> removing in November 2017 already? :/

What would you propose?
Btw I couldn't find any formal deprecation guidelines, but there's some movement to formalize deprecation for [DMD](https://trello.com/c/nfgTwXw6/283-deprecation-guidelines).

jpf91 commented:
LGTM though I don't know anything about the phobos deprecation process. Do we have any guidelines whether a package should import all submodules? I.e. will users expect access to the `std.digest.md5` members when importing `std.digest`?

JackStouffer commented:
I'm not really sold on the idea of breaking people's code for such a cosmetic change.

schveiguy commented:
> I'm not really sold on the idea of breaking people's code for such a cosmetic change.

This doesn't really "Break" people's code, but it does break anyone who does not use the -d switch (a lot of people) without modifying their build.

@wilzbach can you change the deprecation to a `pragma(msg, "deprecation...")` statement?

And the cosmetic change is minor, while the current code has a "cobweb" feel of being based on old restrictions. I doubt that even without the package.d feature, std.digest.digest module (with digest function) would be allowed with the scrutiny we have now.

> removing in November 2017 already? :/

CC @jmdavis who has very strong opinions on deprecation timetables :) I'm thinking November 2017 is not too quick, as that will likely be at least 3 full releases. I don't really like date-based deprecation timetables, as it's really releases that drive when your code breaks or doesn't.

schveiguy commented:
BTW, thanks for making this, @wilzbach!

schveiguy commented:
> Do we have any guidelines whether a package should import all submodules?

No, and there are examples of both in phobos. e.g. std.range imports everything under the package, whereas std.experimental.allocator does not.

schveiguy commented:
@wilzbach you need to update the plethora of windows mak files.

andralex commented:
Could you please keep digest.d and also package.d? That way we introduce the better name without breaking code. We can then phase off digest.d in documentation and later deprecate the module.

wilzbach commented:
> @wilzbach you need to update the plethora of windows mak files.

Ah, thanks (done) - we really need to get rid of these.

>> @wilzbach can you change the deprecation to a pragma(msg, "deprecation...") statement?
> Could you please keep digest.d and also package.d?

I just removed the deprecation message for now. However, `std.digest.digest` will vanish from the documentation (in favor of `std.digest`), but no code will break.

I also added a manual changelog entry that explains the motivation to the user as this is probably easier to see than in the list of fixed Phobos bugs.
(hence the rebase)

The full-text changelog entry should also ensure that people googling for `std.digest.digest` should find something easily and we can avoid having the weird status quo as with [std.typetuple](https://dlang.org/phobos/std_typetuple.html).

andralex commented:
>However, std.digest.digest will vanish from the documentation (in favor of std.digest), but no code will break.

@MartinNowak I'm in favor of experimenting with this approach (have things silently go away from the documentation and get actually deprecated later). The only downside I can think of is people see some old code, try to find it in the docs, and get confused. One possibility is to insert a link to the appropriate changelog entry with a tag such as "See obsolete name `std.digest.digest.digest` in old code?".

wilzbach commented:
> @MartinNowak I'm in favor of experimenting with this approach (have things silently go away from the documentation and get actually deprecated later). The only downside I can think of is people see some old code, try to find it in the docs, and get confused. One possibility is to insert a link to the appropriate changelog entry with a tag such as "See obsolete name std.digest.digest.digest in old code?".

@MartinNowak any reason not to go this way? (this PR is waiting for your opinion as release tzar)

wilzbach commented:
> @MartinNowak I'm in favor of experimenting with this approach (have things silently go away from the documentation and get actually deprecated later).

@andralex @jmdavis @JackStouffer how do we proceed on this? This has been hanging in the queue for quite some time and solely depends on a decision to go with this approach.

JackStouffer commented on std/digest/digest.d:
```diff
@@ -1,1171 +1,3 @@
-        assert(!secureEqual(hex1, hex2));
-    }
-}
+public import std.digest;
```
Is it possible to deprecate this public import with a future removal date (preferably 18 months rather than 12).

jmdavis replied:
I just tried deprecating a public import, and it has no effect. So, it doesn't look like it's possible to deprecate public imports. Now, what we _could_ do (though it would be a bit messy) would be to change the import to a normal import and then create explicit aliases to all of the symbols that were previously, publicly imported. We could then deprecate the aliases.

JackStouffer replied:
Does creating an empty `shared static this` module constructor with a deprecation work?

wilzbach commented on std/digest/digest.d:
```diff
@@ -1,1171 +1,50 @@
+
+// Explicitly undocumented. It will be removed in December 2018. @@@DEPRECATED_2018-12@@@
+deprecated("Please use std.digest instead.")
+alias isDigest = std.digest.isDigest;
```
> I just tried deprecating a public import, and it has no effect. So, it doesn't look like it's possible to deprecate public imports. Now, what we could do (though it would be a bit messy) would be to change the import to a normal import and then create explicit aliases to all of the symbols that were previously, publicly imported. We could then deprecate the aliases.

Nice idea! I gave this a try and except for `digest` it would work nicely.

JackStouffer replied:
Can you add the removal date to the message?

wilzbach commented:
> Can you add the removal date to the message?

Sure, but this still depends on approval.
@MartinNowak are you okay with this?
@jmdavis December 2018 should more than enough time for a deprecation?

wilzbach commented:
> I'm in favor of experimenting with this approach (have things silently go away from the documentation and get actually deprecated later). 

I'm happy to remove the `deprecation` triggers if it unblocks this PR. Otherwise I don't see anything what I could do from my side.

jmdavis commented:
If the deprecations are there, is it possible for someone to write their code so that it doesn't have any deprecation messages with these changes and still compiles with the last release? I _think_ that the answer is no, but it's been a while since I looked at this. If it's not possible, please remove the deprecations, and we'll wait a release to add them. While we don't always manage it, we try to make it so that it's possible to build your code with both the latest release and master without deprecation messages.

wilzbach commented:
> If the deprecations are there, is it possible for someone to write their code so that it doesn't have any deprecation messages with these changes and still compiles with the last release? 

Why not? Everything is now in `std.digest` and `std.digest.digest` just contains the deprecated symbols. Phobos is compiled with `-de`, so any use of `std.digest.digest` within Phobos, which could lead to deprecation messages showing up for a user, is prevented.

> If it's not possible, please remove the deprecations, and we'll wait a release to add them.

As this seems to be the controversial bit about this PR, I commented the deprecation warnings, s.t. it's easy to trigger them once we are ready & we finally can move ahead here.

jmdavis commented:
> Why not? Everything is now in std.digest and std.digest.digest just contains the deprecated symbols. 

If you imported std.digest.digest before, and the deprecation messages are there, you're going to get deprecation messages now, right? And to fix that, you'd import std.digest? If so, wouldn't your code then not compile with the last release, because std.digest.package was added with this PR?

wilzbach commented:
> If you imported std.digest.digest before, and the deprecation messages are there, you're going to get deprecation messages now, right? And to fix that, you'd import std.digest? If so, wouldn't your code then not compile with the last release, because std.digest.package was added with this PR?

AFAICT it would compile fine - one just would get a couple of deprecation warnings, but that's why they are warnings and not compile errors.
Do you mean that it should always be possible for people to compile their project without any deprecation warnings (would be a new policy to me)?
However, I am not so sure whether many people will read the changelog and proactively update their code, but I have already removed/commented the deprecation warnings, s.t. we don't need to discuss this further here, but in general, as mentioned in another PR, a deprecation "policy" for Phobos would be really appreciated ;-)

jmdavis commented:
> Do you mean that it should always be possible for people to compile their project without any deprecation warnings (would be a new policy to me)?

That's exactly what I mean. It should be possible to build the latest release and master without getting deprecation messages - either because no deprecation message has been added yet, or because it's possible to update your code so that it does not get deprecation messages with master and yet still compiles and works with the latest release.

The prime example of where you can run into this sort of problem is adding a new symbol to replace the old one. If we add the new symbol and deprecate the old one at the same time, then you either have deprecation messages in your code when you build with master (because you're using the old symbol), or you've updated your code to use the new symbol and it no longer compiles with the latest release. By adding the new symbol in one release and adding the deprecation message in the next, we avoid that problem.

We don't always manage to follow this policy, but I try to make it so that we do. It started because Vladimir was getting a ton of deprecation messages because of a change, and he couldn't fix it and have his stuff build on both the latest and master. And by having such a policy, it makes it so that library maintainers in general can make their code will work without deprecation messages, whereas otherwise, they're forced to either put up with deprecation messages or not bother ensuring that their code works with master prior to the next release. And while sometimes a deprecation message is only going to pop up once or twice and thus not be a big deal, sometimes it can pop up a _ton_ just because it was used in a template. So, this policy can be a definite quality of life improvement for some developers. This policy is why the old benchmarking functions in std.datetime are documented as scheduled for deprecation in 2.076 rather than already being deprecated (the new functions are being added with 2.075).

jmdavis commented:
Remove the commented out deprecation messages (I think that it will work to do a single one as a label later, but we can leave that until later) and fix the merge conflict, and I think that this is good to go. We can do the actual deprecation for 2.077 (since these changes should go out with 2.076).

wilzbach commented:
> Remove the commented out deprecation messages (I think that it will work to do a single one as a label later, but we can leave that until later) and fix the merge conflict, and I think that this is good to go.

Done (sorry for the rebasing noise - I realized that I can squash the commits).

> (since these changes should go out with 2.076).

Yes the master/stable branch-off for 2.075 has happened four weeks ago: https://github.com/dlang/phobos/pull/5491

jmdavis commented on std/digest/digest.d:
```diff
@@ -1,1171 +1,21 @@
+static import std.digest;
+
+// scheduled for deprecation in 2.077
+// See also: https://github.com/dlang/phobos/pull/5013#issuecomment-313987845
```
I thought that there was a way to just start a review and not add any comments to a line of code, but I don't see a way to do that now. :(

So, here's a useless comment.

schveiguy replied:
Select the "files" tab, then at the top, there is a "Review Changes" button. You can approve the changes without any comments.

BTW, thanks for getting this in!

CyberShadow commented:
@wilzbach Could you have a look at https://issues.dlang.org/show_bug.cgi?id=17724 - was that expected & WONTFIX or an accidental regression?

schveiguy commented:
@CyberShadow it wasn't expected, but was left to try not to break old code. See Andrei's comment above about having both package.d and digest.d. So Does std.digest.digest module work at all? Should we just remove it?

CyberShadow commented:
> So Does std.digest.digest module work at all?

I think the problem is just with the `digest` symbol in `std.digest.digest`.

schveiguy commented:
I'm looking at how this was done, and it's weird.

I tried changing the entire std.digest.digest module to just public import std.digest, and it seems to fix the problem.

It appears that the digest function isn't usable with the way these aliases are done. We should just do it the normal way, and not worry about deprecation. Maybe a pragma(msg) for the whole module instead.

schveiguy commented:
Alternatively, you can alias everything the way it's done now, except for std.digest.digest, and just import that via:

```D
public import std.digest: digest;
```

schveiguy commented:
Figured out a better way to fix it. PR forthcoming.

schveiguy commented:
@CyberShadow #5674
